langkit/support/langkit_support-lexical_env.adb

```diff
@@ -574,11 +575,59 @@ package body Langkit_Support.Lexical_Env is
 
       use Internal_Envs;
 
+      procedure Append_All_Nodes
+        (Env : Lexical_Env; From_Rebound : Boolean);
+      --  Add all nodes of the given Env to the results. Make sure the output
+      --  is deterministic using a sorting procedure.
+
       function Get_Nodes
         (Env : Lexical_Env; From_Rebound : Boolean := False) return Boolean;
       --  Lookup for matching nodes in Env's internal map and append them to
       --  Local_Results. Return whether we found some.
 
+      ----------------------
+      -- Append_All_Nodes --
+      ----------------------
+
+      procedure Append_All_Nodes
+        (Env : Lexical_Env; From_Rebound : Boolean)
+      is
+         type Cursor_Array is array
+           (Natural range <>) of Internal_Envs.Cursor;
+
+         function "<" (A, B : Internal_Envs.Cursor) return Boolean is
+           (Internal_Envs.Key (A).all < Internal_Envs.Key (B).all);
```

Comment:
There is still nondeterminism when we have multiple entries with the same keys, right? At this point I wonder: maybe we should define a total order for nodes just for internal use (it has no special meaning: we just want deterministic order) that would order on 1) the unit filename 2) the node sloc_range. What do you think?

Reply:
Mhh I'm not sure what you mean. The indeterminism I am trying to fix here is the indeterminism that is inherent to the implementation of a hashtable in GNAT. But those are not "multi-hashmaps" so "multiple entries with the same keys" is not possible. I guess you are wondering what happens when we add several nodes with the same key (say, K) in the same env (say, E). Well, in that case the order is well defined, and it's the order in which populate lexical env operates (which AFAIK is deterministic), which will append the vector associated to key K in env E with the new node each time the add_to_env action is executed.

Reply:
However you made me realize that it would be a good idea to test the addition of multiple nodes with the same key, so I will update my testcase accordingly :)

Reply:
FWIW my understanding is the same as Romain's, ie. there is no indeterminism when you have multiple entries with the same key: If they're in the same env, they're order (by insertion order). If they're in different envs, they're still ordered (by traversal order).